server/cluster.go

```diff
@@ -364,41 +364,72 @@ func (c *RaftCluster) GetStore(storeID uint64) (*metapb.Store, *StoreStatus, err
 }
 
 func (c *RaftCluster) putStore(store *metapb.Store) error {
+	c.Lock()
+	defer c.Unlock()
+
 	if store.GetId() == 0 {
 		return errors.Errorf("invalid put store %v", store)
 	}
 
-	storeValue, err := store.Marshal()
-	if err != nil {
-		return errors.Trace(err)
+	// There are 3 cases here:
+	// Case 1: store id exists with the same address - do nothing;
+	// Case 2: store id exists with different address - update address;
+	if s := c.cachedCluster.getStore(store.GetId()); s != nil {
+		if s.store.GetAddress() == store.GetAddress() {
+			return nil
+		}
+		s.store.Address = store.Address
+		return c.saveStore(s.store)
 	}
 
-	storePath := makeStoreKey(c.clusterRoot, store.GetId())
-
-	// Fix #287, check duplicated store address first.
-	stores := c.cachedCluster.getStores()
-	for _, s := range stores {
+	// Case 3: store id does not exist, check duplicated address.
+	for _, s := range c.cachedCluster.getStores() {
 		if s.store.GetAddress() == store.GetAddress() {
-			if s.store.GetId() == store.GetId() {
-				return nil
-			}
 			return errors.Errorf("duplicated store address: %s, already registered by id: %d", store.GetAddress(), store.GetId())
```

Comment:
Seems this log`already registered by id` should be `s.store.GetId()`, i also suggest log the new store id.